doc/mgmt/SONiC_DNS_Support.md

```diff
@@ -0,0 +1,433 @@
+# Feature Name
+Domain Name System (DNS) support
+
+# High Level Design Document
+
+#### Rev 0.1
+
+# Table of Contents
+  * [List of Tables](#list-of-tables)
+  * [Revision](#revision)
+  * [About This Manual](#about-this-manual)
+  * [Scope](#scope)
+  * [Definition/Abbreviation](#definitionabbreviation)
+
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+# Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 | 05/05 |   Venkatesan Mahalingam         | Initial version                   |
+| 0.2 | 05/11 |   Venkatesan Mahalingam         | Added source interface support and addressed comments                   |
+
+
+# About this Manual
+
+This document introduces the support of DNS in SONiC.
+
+# Scope
+
+This document covers the following,
+1) Click commands to configure DNS server & Source interface
+2) Click command to show DNS hosts
+3) New tables in config DB to handle DNS configs
+4) DNS table configuration and show based on OpenConfig YANG model
+5) KLISH commands to configure DNS server & Source interface
+6) KLISH commands to show DNS hosts
+7) Backend support to add DNS configs in "/etc/resolv.conf" file & iptable rules to change the source-IP of the DNS query packets
+8) Unit Testcases
+
+# Definition/Abbreviation
+
+### Table 1: Abbreviations
+| **Term**                 | **Meaning**                         |
+|--------------------------|-------------------------------------|
+|          DNS             |  Domain Name System             |
+
+
+
+# 1 Feature Overview
+
+DNS stands for Domain Name System. It is used to translate human readable domain names into machine readable IP addresses.
+
+With this feature, users will be able to configure DNS servers and source interface using SONiC Click commands and north bound interface (KLISH/REST/gNMI) provided by management framework module using OpenConfig models.
+
+Also, backend support to handle new config DB table events to populate nameservers in "/etc/resolv.conf" and add iptable rules to change the source IP (present on the source interface) of the DNS query that is being sent to the DNS server.

```

Comment:
Do we take the primary address? Can we distinguish the IP alias from the main address of the interface to use? What is the difference between source-interface and source-address? Are we supporting both?

Reply:
We dont have any flag to indicate primary address in the config-DB interface tables, it's hard to get the primary address from config-DB table unless we go directly read the IP address with flag from kernel (but there can be many primary addresses, see below kernel output), for now, we can add the caveat that when there are more than one IP addresses on the interface, we pick the address based on first IP address that we get in the table read, so having only one IPv4/IPv6 address on source interface is recommended in order to make the behavior deterministic. We support only source-interface now not source-IP any more as per PLM' ask.


root@sonic:~# ip address show Ethernet0
20: Ethernet0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9100 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 90:b1:1c:f4:9d:50 brd ff:ff:ff:ff:ff:ff
    inet 21.0.0.1/16 scope global Ethernet0
       valid_lft forever preferred_lft forever
    inet 22.0.0.1/16 scope global Ethernet0
       valid_lft forever preferred_lft forever
    inet 20.0.0.2/16 scope global secondary Ethernet0
       valid_lft forever preferred_lft forever
    inet 21.0.0.2/16 scope global secondary Ethernet0
       valid_lft forever preferred_lft forever
    inet 22.0.0.2/16 scope global secondary Ethernet0
       valid_lft forever preferred_lft forever
    inet6 2001::1/64 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::92b1:1cff:fef4:9d50/64 scope link flags 4000
       valid_lft forever preferred_lft forever
root@sonic:~# redis-cli -n 4 keys "INTERFACE*"
1) "INTERFACE|Ethernet0|2001::1/64"
2) "INTERFACE|Ethernet0|21.0.0.2/16"
3) "INTERFACE|Ethernet0|21.0.0.1/16"
4) "INTERFACE|Ethernet0|20.0.0.2/16"
5) "INTERFACE|Ethernet0|22.0.0.1/16"
6) "INTERFACE|Ethernet0|20.0.0.1/16"
7) "INTERFACE|Ethernet0"
8) "INTERFACE|Ethernet0|22.0.0.2/16"
root@sonic:~#

Reply:
Yes, please document, since the order of retrieval from the redis-db is redis dependent, and can change between reboot/restart, or in the same session.
Also, what happens if both source-interface and source-ip are configured ?  Seems like overkill, since we are recommending that only one ip address be on the interface. Additionally the openconfig-system.yang already has a definition for ntp-source-address? Why to define another way to set the ntp source where it is not even deterministic assignment of ip address?

Reply:
We discussed this with PLM, interface name is preferred over IP address to avoid any IP config errors, we dont support source-IP configuration.